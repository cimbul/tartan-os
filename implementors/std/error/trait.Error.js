(function() {var implementors = {
"aml_parse":[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/nightly/std/error/trait.Error.html\" title=\"trait std::error::Error\">Error</a> for <a class=\"struct\" href=\"aml_parse/struct.UsageError.html\" title=\"struct aml_parse::UsageError\">UsageError</a>"]]
};if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()