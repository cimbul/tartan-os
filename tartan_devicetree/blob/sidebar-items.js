window.SIDEBAR_ITEMS = {"enum":[["StructureData","An event that signals the start/end of a node or a property of one while walking the device tree."]],"struct":[["Tree","Represents a full flattened device tree blob (DTB)."]]};