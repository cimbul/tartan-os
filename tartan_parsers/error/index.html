<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Helpers for reporting parsing errors"><title>tartan_parsers::error - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-e935ef01ae1c1829.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="tartan_parsers" data-themes="" data-resource-suffix="" data-rustdoc-version="1.78.0-nightly (381d69953 2024-02-24)" data-channel="nightly" data-search-js="search-dd67cee4cfa65049.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../static.files/storage-4c98445ec4002617.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-c37d3936c59ababd.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-04d5337699b92874.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../tartan_parsers/index.html">tartan_parsers</a><span class="version">0.1.0</span></h2></div><h2 class="location"><a href="#">Module error</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li></ul></section><h2><a href="../index.html">In crate tartan_parsers</a></h2></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../../tartan_parsers/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../index.html">tartan_parsers</a>::<wbr><a class="mod" href="#">error</a><button id="copy-path" title="Copy item path to clipboard"><img src="../../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../../src/tartan_parsers/error.rs.html#1-135">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Helpers for reporting parsing errors</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.ErrorWithPosition.html" title="struct tartan_parsers::error::ErrorWithPosition">ErrorWithPosition</a></div><div class="desc docblock-short">A parser error with position info useful for printing human-readable messages.</div></li><li><div class="item-name"><a class="struct" href="struct.Position.html" title="struct tartan_parsers::error::Position">Position</a></div><div class="desc docblock-short">Helper struct that prints human-readable position information when formatted with
<code>Display</code>.</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.GeneralParseError.html" title="trait tartan_parsers::error::GeneralParseError">GeneralParseError</a></div><div class="desc docblock-short">Trait alias combining the nom error traits [<code>ParseError</code>] and [<code>ContextError</code>].</div></li></ul><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.err.html" title="fn tartan_parsers::error::err">err</a></div><div class="desc docblock-short">Shortcut for constructing parser errors</div></li></ul></section></div></main></body></html>