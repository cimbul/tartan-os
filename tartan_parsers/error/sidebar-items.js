window.SIDEBAR_ITEMS = {"fn":["err"],"struct":["ErrorWithPosition","Position"],"trait":["GeneralParseError"]};