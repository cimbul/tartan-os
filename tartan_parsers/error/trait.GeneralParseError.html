<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Trait alias combining the nom error traits `ParseError` and `ContextError`."><meta name="keywords" content="rust, rustlang, rust-lang, GeneralParseError"><title>GeneralParseError in tartan_parsers::error - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceSerif4-Regular-1f7d512b176f0f72.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceSerif4-Bold-124a1ca42af929b6.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-1cc40db21e31a224.css" id="mainThemeStyle"><link rel="stylesheet" id="themeStyle" href="../../static.files/light-8fe2cf7ee64a76fb.css"><link rel="stylesheet" disabled href="../../static.files/dark-7aacebe982779e9a.css"><link rel="stylesheet" disabled href="../../static.files/ayu-29dcc614f7a7b461.css"><script id="default-settings" ></script><script src="../../static.files/storage-d43fa987303ecbbb.js"></script><script defer src="sidebar-items.js"></script><script defer src="../../static.files/main-a211dbb005fb8161.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-13285aec31fa243e.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc trait"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../../tartan_parsers/index.html"><div class="logo-container"><img class="rust-logo" src="../../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></div></a><h2></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../../tartan_parsers/index.html"><div class="logo-container"><img class="rust-logo" src="../../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></div></a><h2 class="location"><a href="#">GeneralParseError</a></h2><div class="sidebar-elems"><section><h3><a href="#implementors">Implementors</a></h3></section><h2><a href="index.html">In tartan_parsers::error</a></h2></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../static.files/wheel-5ec35bf9ca753509.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn">Trait <a href="../index.html">tartan_parsers</a>::<wbr><a href="index.html">error</a>::<wbr><a class="trait" href="#">GeneralParseError</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../../src/tartan_parsers/error.rs.html#10-14">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><div class="item-decl"><pre class="rust trait"><code>pub trait GeneralParseError&lt;I&gt;<span class="where fmt-newline">where<br>&nbsp;&nbsp;&nbsp;&nbsp;Self: ParseError&lt;I&gt; + ContextError&lt;I&gt;,</span>{ }</code></pre></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Trait alias combining the nom error traits <a href="nom::error::ParseError"><code>ParseError</code></a> and
<a href="nom::error::ContextError"><code>ContextError</code></a>.</p>
</div></details><h2 id="implementors" class="small-section-header">Implementors<a href="#implementors" class="anchor">§</a></h2><div id="implementors-list"><section id="impl-GeneralParseError%3CI%3E-for-T" class="impl has-srclink"><a class="srclink rightside" href="../../src/tartan_parsers/error.rs.html#16">source</a><a href="#impl-GeneralParseError%3CI%3E-for-T" class="anchor">§</a><h3 class="code-header">impl&lt;T, I&gt; <a class="trait" href="trait.GeneralParseError.html" title="trait tartan_parsers::error::GeneralParseError">GeneralParseError</a>&lt;I&gt; for T<span class="where fmt-newline">where<br>&nbsp;&nbsp;&nbsp;&nbsp;T: ParseError&lt;I&gt; + ContextError&lt;I&gt;,</span></h3></section></div><script src="../../implementors/tartan_parsers/error/trait.GeneralParseError.js" async></script></section></div></main><div id="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="tartan_parsers" data-themes="" data-resource-suffix="" data-rustdoc-version="1.68.0-nightly (9c07efe84 2022-12-16)" data-search-js="search-444266647c4dba98.js" data-settings-js="settings-bebeae96e00e4617.js" data-settings-css="settings-58836c674e2f7bd2.css" ></div></body></html>