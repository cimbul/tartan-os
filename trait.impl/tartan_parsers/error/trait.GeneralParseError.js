(function() {
    var implementors = Object.fromEntries([["tartan_parsers",[]]]);
    if (window.register_implementors) {
        window.register_implementors(implementors);
    } else {
        window.pending_implementors = implementors;
    }
})()
//{"start":57,"fragment_lengths":[21]}