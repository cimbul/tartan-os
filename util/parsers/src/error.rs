use core::cmp::min;
use core::fmt;
use nom::error::{ContextError, ErrorKind, ParseError, VerboseError, VerboseErrorKind};
use nom::{AsBytes, IResult};


/// Trait alias combining the nom error traits [`ParseError`](nom::error::ParseError) and
/// [`ContextError`](nom::error::ContextError).
#[allow(clippy::module_name_repetitions)]
pub trait GeneralParseError<I>
where
    Self: ParseError<I> + ContextError<I>,
{
}

impl<T, I> GeneralParseError<I> for T where T: ParseError<I> + ContextError<I> {}


/// Shortcut for constructing parser errors
#[allow(clippy::missing_errors_doc)]
pub fn err<I, O, E: ParseError<I>>(i: I, kind: ErrorKind) -> IResult<I, O, E> {
    Err(nom::Err::Error(E::from_error_kind(i, kind)))
}


/// Helper struct that prints human-readable position information when formatted with
/// `Display`.
pub struct Position<'a, I: AsBytes = &'a [u8]> {
    /// The state of the parser at the time we are interested in
    pub state: &'a I,
    /// The complete input we are trying to parse
    pub full_input: &'a [u8],
}

impl<I: AsBytes> fmt::Display for Position<'_, I> {
    fn fmt(&self, out: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
        if self.full_input.is_empty() {
            return writeln!(out, "in empty input");
        }

        // NOTE: nom's Offset implementation can panic. Avoid it.
        let full_input_pos = self.full_input.as_ptr() as usize;
        let state_pos = self.state.as_bytes().as_ptr() as usize;
        let offset = match state_pos.checked_sub(full_input_pos) {
            Some(o) => o,
            None => return writeln!(out, "at unknown/invalid offset"),
        };

        let context_start = offset.saturating_sub(10);
        let context_end = min(offset + 10, self.full_input.len());
        let context_slice = &self.full_input[context_start..context_end];

        write!(out, "at offset {offset} ({offset:#x}):\n ")?;
        for byte in context_slice {
            let ascii_char = match byte {
                b if b.is_ascii_graphic() => *b as char,
                b' ' => ' ',
                _ => '.',
            };
            write!(out, " {ascii_char:2}")?;
        }
        write!(out, "\n ")?;
        for byte in context_slice {
            write!(out, " {byte:02x}")?;
        }
        write!(
            out,
            "\n  {caret:>caret_offset$}\n",
            caret = '^',
            caret_offset = (offset - context_start) * 3 + 1,
        )
    }
}


/// A parser error with position info useful for printing human-readable messages.
#[derive(Debug, Clone)]
#[allow(clippy::module_name_repetitions)]
pub struct ErrorWithPosition<'a, I = &'a [u8]>
where
    I: From<&'a [u8]> + AsBytes,
{
    /// The error returned by the parser
    pub error: VerboseError<I>,
    /// The complete input we were trying to parse
    pub full_input: &'a [u8],
}

impl<'a, I> ErrorWithPosition<'a, I>
where
    I: From<&'a [u8]> + AsBytes,
{
    /// Associate a parse error with its full input, to reference the affected position
    /// in error messages.
    pub fn new(wrapped_error: nom::Err<VerboseError<I>>, full_input: &'a [u8]) -> Self {
        match wrapped_error {
            nom::Err::Error(error) | nom::Err::Failure(error) => {
                ErrorWithPosition { error, full_input }
            }
            nom::Err::Incomplete(_) => {
                let error = VerboseError::from_error_kind(
                    I::from(full_input),
                    ErrorKind::Complete,
                );
                ErrorWithPosition { error, full_input }
            }
        }
    }
}

impl<'a, I> fmt::Display for ErrorWithPosition<'a, I>
where
    I: From<&'a [u8]> + AsBytes,
{
    fn fmt(&self, out: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
        if self.error.errors.is_empty() {
            return writeln!(out, "Unknown parsing error");
        }

        for (state, kind) in &self.error.errors {
            match kind {
                VerboseErrorKind::Context(context) => {
                    write!(out, "In {context}")?;
                }
                VerboseErrorKind::Char(c) => {
                    write!(out, "Expected '{c}'")?;
                }
                VerboseErrorKind::Nom(e) => {
                    write!(out, "Failed {}", e.description())?;
                }
            }
            write!(out, " {}", Position { state, full_input: self.full_input })?;
        }

        Ok(())
    }
}
