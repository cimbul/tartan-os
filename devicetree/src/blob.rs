//! Flattened device tree, a.k.a. devicetree blob (DTB), as defined in chapter 5 of the
//! Devicetree spec.

use crate::{MemoryReservation, Value};
use core::mem::{size_of, size_of_val};
use core::slice;
use core::str;
use nom::branch::alt;
use nom::bytes::complete as bytes;
use nom::combinator::{map, map_parser, value, verify};
use nom::error::{ErrorKind, ParseError};
use nom::number::complete as num;
use nom::sequence::{preceded, terminated};
use nom::IResult;
use tartan_parsers::error::{err, GeneralParseError};
use tartan_parsers::{opcode, result_iterator, struct_parser};


/// Represents a full flattened device tree blob (DTB).
#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub struct Tree<'a> {
    header: Header,
    memory_reservation_block: &'a [u8],
    structure_block: &'a [u8],
    string_block: &'a [u8],
}

impl<'a> Tree<'a> {
    /// Parse a devicetree blob that begins at the specified memory address.
    ///
    /// # Safety
    /// The pointer must hold the address of the first byte of a valid devicetree blob,
    /// and it must uphold all the invariants of [`core::slice::from_raw_parts`] for a
    /// slice beginning at the specified address and extending the total length of the DTB
    /// as given in its header.
    ///
    /// # Errors
    /// If the header does not have the correct magic number, or any other properties of
    /// the header are invalid, this function will return an error.
    pub unsafe fn from_ptr<E: ParseError<&'a [u8]>>(
        ptr: *const u8,
    ) -> Result<Self, nom::Err<E>> {
        let header_data = slice::from_raw_parts(ptr, Header::SIZE);
        let (_, header) = Header::parse(header_data)?;

        let full_data = slice::from_raw_parts(ptr, header.total_size as usize);

        Ok(Self::from_buffer_and_header(full_data, header))
    }


    /// Parse a devicetree blob that starts at the beginning of the slice and extends no
    /// further than its end.
    ///
    /// # Errors
    /// If the header does not have the correct magic number, or any other properties of
    /// the header are invalid, this function will return an error.
    pub fn from_buffer<E: ParseError<&'a [u8]>>(
        data: &'a [u8],
    ) -> Result<Self, nom::Err<E>> {
        let (_, header) = Header::parse(data)?;
        Ok(Self::from_buffer_and_header(data, header))
    }


    fn from_buffer_and_header(data: &'a [u8], header: Header) -> Self {
        let mut start: usize;
        let mut len: usize;

        start = header.memory_reservation_block_offset.try_into().unwrap();
        let end = header.structure_block_offset.try_into().unwrap();
        let memory_reservation_block = &data[start..end];

        start = header.structure_block_offset.try_into().unwrap();
        len = header.structure_block_size.try_into().unwrap();
        let structure_block = &data[start..start + len];

        start = header.string_block_offset.try_into().unwrap();
        len = header.string_block_size.try_into().unwrap();
        let string_block = &data[start..start + len];

        Self { header, memory_reservation_block, structure_block, string_block }
    }


    /// Iterate over memory reservations defined by this devicetree
    pub fn memory_reservation_iter<E: ParseError<&'a [u8]> + 'a>(
        self: &'a Tree<'a>,
    ) -> impl Iterator<Item = Result<MemoryReservation, nom::Err<E>>> + 'a {
        result_iterator(self.memory_reservation_block, MemoryReservation::parse)
    }


    /// Iterate over elements in the device tree. Note that this does not directly iterate
    /// over *nodes*, but rather events that indicate the start of a node (including its
    /// children), a property defined on the node, or the end of a node.
    pub fn structure_iter<E: GeneralParseError<&'a [u8]>>(
        self: &'a Tree<'a>,
    ) -> impl Iterator<Item = Result<StructureData<'a>, nom::Err<E>>> {
        result_iterator(self.structure_block, StructureToken::parse)
            .take_while(|s| !matches!(s, Ok(StructureToken::End)))
            .filter_map(move |s| match s {
                Ok(StructureToken::BeginNode(n)) => Some(Ok(StructureData::BeginNode(n))),
                Ok(StructureToken::EndNode) => Some(Ok(StructureData::EndNode)),
                Ok(StructureToken::Property { name_offset, value }) => {
                    let value = Value { data: value };
                    match self.get_string(name_offset) {
                        Err(e) => Some(Err(e)),
                        Ok(name) => Some(Ok(StructureData::Property { name, value })),
                    }
                }
                Ok(StructureToken::NoOp | StructureToken::End) => None,
                Err(e) => Some(Err(e)),
            })
    }


    /// Retrieve a null-terminated string at the given offset in the strings block
    fn get_string<E: GeneralParseError<&'a [u8]>>(
        self: &Tree<'a>,
        offset: u32,
    ) -> Result<&'a str, nom::Err<E>> {
        let offset = offset as usize;
        let (_, string) = parse_c_string(&self.string_block[offset..])?;
        Ok(string)
    }
}


#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
struct Header {
    pub total_size: u32,
    pub structure_block_offset: u32,
    pub string_block_offset: u32,
    pub memory_reservation_block_offset: u32,
    pub version: u32,
    pub last_compatible_version: u32,
    pub boot_cpu_id: u32,
    pub string_block_size: u32,
    pub structure_block_size: u32,
}

impl Header {
    /// Expected size of serialized header
    pub const SIZE: usize = size_of_val(&Self::MAGIC) + size_of::<Self>();

    const MAGIC: [u8; 4] = [0xd0, 0x0d, 0xfe, 0xed];
    const VERSION: u32 = 17;

    pub fn parse<'a, E: ParseError<&'a [u8]>>(i: &'a [u8]) -> IResult<&'a [u8], Self, E> {
        preceded(
            bytes::tag(Self::MAGIC),
            verify(
                struct_parser! {
                    Self {
                        total_size: num::be_u32,
                        structure_block_offset: num::be_u32,
                        string_block_offset: num::be_u32,
                        memory_reservation_block_offset: num::be_u32,
                        version: num::be_u32,
                        last_compatible_version: num::be_u32,
                        boot_cpu_id: num::be_u32,
                        string_block_size: num::be_u32,
                        structure_block_size: num::be_u32,
                    }
                },
                |h| h.last_compatible_version <= Self::VERSION,
            ),
        )(i)
    }
}


impl MemoryReservation {
    fn parse<'a, E: ParseError<&'a [u8]>>(i: &'a [u8]) -> IResult<&'a [u8], Self, E> {
        let parser = struct_parser! {
            Self {
                address: num::be_u64,
                size: num::be_u64,
            }
        };
        parser(i)
    }
}


/// An event that signals the start/end of a node or a property of one while walking the
/// device tree.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StructureData<'a> {
    /// Indicates that the following events are within the scope of a new node defined
    /// as the child of the previous current node.
    BeginNode(&'a str),
    /// Indicates that the previous current node is complete, and should be popped off
    /// the stack and replaced with its parent.
    EndNode,
    /// Defines a property of the current node
    Property {
        /// The name of the property.
        ///
        /// The Devicetree spec lists a number of common property names, but the system
        /// is open for extension, so it may not be recognizable.
        name: &'a str,

        /// The value of the property.
        ///
        /// The format of this data depends on the property name. Types defined by the
        /// Devicetree spec include:
        ///   * empty values,
        ///   * null-terminated strings or lists of them,
        ///   * big-endian 32-bit or 64-bit integers
        ///   * "phandles," or 32-bit numbers that uniquely identify another node
        ///
        /// However, different properties can define their own representation, so it is
        /// not possible to infer the type of an unrecognized property.
        value: Value<'a>,
    },
}


#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum StructureToken<'a> {
    BeginNode(&'a str),
    EndNode,
    Property { name_offset: u32, value: &'a [u8] },
    NoOp,
    End,
}

impl<'a> StructureToken<'a> {
    fn parse<E: GeneralParseError<&'a [u8]>>(i: &'a [u8]) -> IResult<&'a [u8], Self, E> {
        alt((
            Self::parse_begin_node,
            Self::parse_end_node,
            Self::parse_property,
            Self::parse_no_op,
            Self::parse_end,
        ))(i)
    }

    fn parse_begin_node<E: GeneralParseError<&'a [u8]>>(
        i: &'a [u8],
    ) -> IResult<&'a [u8], Self, E> {
        opcode(
            "begin node",
            bytes::tag(1_u32.to_be_bytes()),
            map(alignment_padded(4, parse_c_string), Self::BeginNode),
        )(i)
    }

    fn parse_end_node<E: ParseError<&'a [u8]>>(
        i: &'a [u8],
    ) -> IResult<&'a [u8], Self, E> {
        value(Self::EndNode, bytes::tag(2_u32.to_be_bytes()))(i)
    }

    fn parse_property<E: GeneralParseError<&'a [u8]>>(
        i: &'a [u8],
    ) -> IResult<&'a [u8], Self, E> {
        opcode(
            "property",
            bytes::tag(3_u32.to_be_bytes()),
            alignment_padded(4, |i| {
                let (i, value_length) = num::be_u32(i)?;
                let (i, name_offset) = num::be_u32(i)?;
                let (i, value) = bytes::take(value_length)(i)?;
                Ok((i, Self::Property { name_offset, value }))
            }),
        )(i)
    }

    fn parse_no_op<E: ParseError<&'a [u8]>>(i: &'a [u8]) -> IResult<&'a [u8], Self, E> {
        value(Self::NoOp, bytes::tag(4_u32.to_be_bytes()))(i)
    }

    fn parse_end<E: ParseError<&'a [u8]>>(i: &'a [u8]) -> IResult<&'a [u8], Self, E> {
        value(Self::End, bytes::tag(9_u32.to_be_bytes()))(i)
    }
}


fn parse_c_string<'a, E: GeneralParseError<&'a [u8]>>(
    i: &'a [u8],
) -> IResult<&'a [u8], &'a str, E> {
    let null_terminated_bytes =
        terminated(bytes::take_until(&[0_u8] as &[u8]), bytes::tag([0_u8]));
    map_parser(null_terminated_bytes, |b| match str::from_utf8(b) {
        Ok(s) => Ok((&[] as &[u8], s)),
        Err(_) => err(i, ErrorKind::Verify),
    })(i)
}


fn alignment_padded<'a, P, O, E>(
    alignment_bytes: usize,
    parser: P,
) -> impl Fn(&'a [u8]) -> IResult<&'a [u8], O, E>
where
    E: ParseError<&'a [u8]>,
    P: Fn(&'a [u8]) -> IResult<&'a [u8], O, E>,
{
    move |i| {
        let (i, o) = parser(i)?;

        // Skip any bytes necessary to keep the remaining input aligned
        let address = i.as_ptr() as usize;
        let remainder = address % alignment_bytes;
        let padding = if remainder == 0 { 0 } else { alignment_bytes - remainder };
        let (i, _) = bytes::take(padding)(i)?;

        Ok((i, o))
    }
}
