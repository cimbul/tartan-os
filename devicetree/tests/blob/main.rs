#![warn(clippy::pedantic)]

use insta::{assert_debug_snapshot, glob, with_settings};
use nom::error::VerboseError;
use std::fs;
use tartan_devicetree::blob::Tree;

#[test]
fn parse_all() {
    // Parse all DTB files and compare to past output
    glob!("dtb/*.dtb", |dtb_path| {
        // Make the autogenerated snapshot filenames less wordy
        let test_name = dtb_path.file_stem().unwrap().to_str().unwrap();
        with_settings!({snapshot_suffix => ""}, {
            let dtb_data = fs::read(dtb_path).unwrap();
            let tree = Tree::from_buffer::<()>(&dtb_data).unwrap();

            let structures: Vec<_> = tree
                .structure_iter::<VerboseError<&[u8]>>()
                .collect();
            assert_debug_snapshot!(format!("{test_name}_structures"), structures);

            let mem_reservations: Vec<_> = tree
                .memory_reservation_iter::<VerboseError<&[u8]>>()
                .collect();
            assert_debug_snapshot!(format!("{test_name}_mem_resrv"), mem_reservations);
        });
    });
}
