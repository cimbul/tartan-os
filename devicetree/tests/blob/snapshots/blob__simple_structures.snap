---
source: devicetree/tests/blob/main.rs
expression: structures
---
[
    Ok(
        BeginNode(
            "",
        ),
    ),
    Ok(
        BeginNode(
            "foo",
        ),
    ),
    Ok(
        Property {
            name: "bar",
            value: <62617a00> "baz",
        },
    ),
    Ok(
        Property {
            name: "quux",
            value: <00000042>,
        },
    ),
    Ok(
        EndNode,
    ),
    Ok(
        BeginNode(
            "blorg",
        ),
    ),
    Ok(
        Property {
            name: "flarf",
            value: <6e6f7262 6c6500> "norble",
        },
    ),
    Ok(
        EndNode,
    ),
    Ok(
        EndNode,
    ),
]
